src/balmy/multichain-identifier.ts

```diff
@@ -0,0 +1,143 @@
+import { Address, ChainId } from "@mean-finance/sdk";
+import { FullTokenData, TokenData } from "./types";
+import { sortTokens } from "./utils";
+
+type CgChainId = string;
+type CgCoinId = string;
+
+type CoinGeckoCoin = {
+  id: string;
+  symbol: string;
+  name: string;
+  platforms: Record<string, string>;
+};
+
+type CoinGeckoPlatform = {
+  id: CgChainId;
+  chain_identifier: number | null;
+  name: string;
+  shortname: string;
+  native_coin_id: string;
+};
+
+const equivalentCgCoinIds: Record<CgCoinId, Set<CgCoinId>> = {
+  tether: new Set([
+    "bridged-usdt",
+    "arbitrum-bridged-usdt-arbitrum",
+    "bridged-tether-linea",
+    "bridged-tether-manta-pacific",
+    "bridged-tether-opbnb",
+    "bridged-tether-scroll",
+    "bridged-tether-stargate",
+    "bridged-tether-starkgate",
+    "bridged-usdt-core",
+    "celer-bridged-usdt-conflux",
+    "cronos-bridged-usdt-cronos",
+    "gnosis-xdai-bridged-usdt-gnosis",
+    "huobi-bridged-usdt-heco-chain",
+    "kucoin-bridged-usdt-kucoin-community-chain",
+    "lxly-bridged-usdt-astar-zkevm",
+    "mantle-bridged-usdt-mantle",
+    "mapped-usdt",
+    "mode-bridged-usdt-mode",
+    "multichain-bridged-usdt-moonbeam",
+    "multichain-bridged-usdt-moonriver",
+    "multichain-bridged-usdt-syscoin",
+    "multichain-bridged-usdt-telos",
+    "neonpass-bridged-usdt-neon",
+    "nova-tether-usd",
+    "polygon-bridged-usdt-polygon",
+    "polygon-hermez-bridged-usdt-polygon-zkevm",
+    "polygon-hermez-bridged-usdt-x-layer",
+    "shimmerbridge-bridged-usdt-shimmerevm",
+    "symbiosis-bridged-usdt-bahamut",
+    "tether-pulsechain",
+    "thundercore-bridged-usdt-thundercore",
+    "viction-bridged-usdt-viction",
+  ]),
+};
+
+export class MultichainIdentifier {
+  private cgChainMap: Record<CgChainId, ChainId>;
+  private cgMultichainMap: Record<
+    CgCoinId,
+    { chainId: ChainId; address: Address }[]
+  >;
+  private tokenKeyToCgCoinId: Record<string, CgCoinId>;
+
+  constructor() {
+    this.cgChainMap = {};
+    this.cgMultichainMap = {};
+    this.tokenKeyToCgCoinId = {};
+  }
+
+  private async fetchData() {
+    const cgCoinsPromise: Promise<CoinGeckoCoin[]> = fetch(
+      "https://api.coingecko.com/api/v3/coins/list?include_platform=true"
+    ).then((res) => res.json());
+    const cgPlatformsResponse: Promise<CoinGeckoPlatform[]> = fetch(
+      "https://api.coingecko.com/api/v3/asset_platforms"
+    ).then((res) => res.json());
+
+    return await Promise.all([cgCoinsPromise, cgPlatformsResponse]);
+  }
+
+  async prepareData() {
```

Comment:
I think this approach of having `prepareData` and then `populateChainAddresses` can lead to a scenario where someone forgets to call `prepareData` and then `populateChainAddresses` does nothing. A safer approach would be to have a function called `buildMultichainIdentifier(): Promise<MultichainIdentifier>`. This function would fetch data, and then build `MultichainIdentifier` with all data prepared. By doing so, when you call `populateChainAddresses` you are guaranteed to have everything ready. And if you don't build it, you of course can't call `populateChainAddresses`

Reply:
Fantastic! Thank you

Reply:
Agree!